dhall/src/Dhall/DirectoryTree.hs

```diff
@@ -124,18 +248,257 @@ toDirectoryTree path expression = case expression of
         empty
 
     process key value = do
-        if Text.isInfixOf (Text.pack [ FilePath.pathSeparator ]) key
-            then die
-            else return ()
+        when (not allowSeparators && Text.isInfixOf (Text.pack [ FilePath.pathSeparator ]) key) $
+            die
 
-        Directory.createDirectoryIfMissing False path
+        Directory.createDirectoryIfMissing allowSeparators path
 
-        toDirectoryTree (path </> Text.unpack key) value
+        toDirectoryTree allowSeparators (path </> Text.unpack key) value
 
     die = Exception.throwIO FilesystemError{..}
       where
         unexpectedExpression = expression
 
+-- | The type of a fixpoint directory tree expression.
+directoryTreeType :: Expector (Expr Src Void)
+directoryTreeType = Pi Nothing "result" (Const Type)
+    <$> (Pi Nothing "make" <$> makeType <*> pure (App List (Var (V "result" 0))))
+
+-- | The type of make part of a fixpoint directory tree expression.
+makeType :: Expector (Expr Src Void)
+makeType = Record . Map.fromList <$> sequenceA
+    [ makeConstructor "directory" (Decode.auto :: Decoder DirectoryEntry)
+    , makeConstructor "file" (Decode.auto :: Decoder FileEntry)
+    ]
+    where
+        makeConstructor :: Text -> Decoder b -> Expector (Text, RecordField Src Void)
+        makeConstructor name dec = (name,) . Core.makeRecordField
+            <$> (Pi Nothing "_" <$> expected dec <*> pure (Var (V "result" 0)))
+
+type DirectoryEntry = Entry (Seq FilesystemEntry)
+
+type FileEntry = Entry Text
+
+-- | A filesystem entry.
+data FilesystemEntry
+    = DirectoryEntry (Entry (Seq FilesystemEntry))
+    | FileEntry (Entry Text)
+    deriving Show
+
+instance FromDhall FilesystemEntry where
+    autoWith normalizer = Decoder
+        { expected = pure $ Var (V "result" 0)
+        , extract = \case
+            App (Field (Var (V "make" 0)) (fieldSelectionLabel -> "directory")) entry ->
+                DirectoryEntry <$> extract (autoWith normalizer) entry
+            App (Field (Var (V "make" 0)) (fieldSelectionLabel -> "file")) entry ->
+                FileEntry <$> extract (autoWith normalizer) entry
+            expr -> Decode.typeError (expected (Decode.autoWith normalizer :: Decoder FilesystemEntry)) expr
+        }
+
+-- | A generic filesystem entry. This type holds the metadata that apply to all
+-- entries. It is parametric over the content of such an entry.
+data Entry a = Entry
+    { entryName :: String
+    , entryContent :: a
+    , entryUser :: Maybe User
+    , entryGroup :: Maybe Group
+    , entryMode :: Maybe (Mode Maybe)
+    }
+    deriving (Generic, Show)
+
+instance FromDhall a => FromDhall (Entry a) where
+    autoWith = Decode.genericAutoWithInputNormalizer Decode.defaultInterpretOptions
+        { fieldModifier = Text.toLower . Text.drop (Text.length "entry")
```

Comment:
Do we need to prefix the field names with `entry`?

Same question for `mode` and `access` below

Reply:
No, I think this is a leftover. I can change that if you want.

Reply:
Turns out it is no leftover after all: If the prefix gets removed there are ambiguous record fields like `entryUser :: Maybe User` and `modeUser :: f (Access f)`. We could probably work around those but I think it is cleaner to keep the prefix.

Reply:
Alright, it's fine to keep the prefix, then